### DISCUSSION TITLE
Is Quarto meant to support knitr::read_chunk in a book chapter?

### DISCUSSION THREAD

**Discussion by harrelfe** [quarto-dev/quarto-cli#1458] [Q&A]
I have a book chapter in which I have the R code in a separate file `shared.R` in `knitr::read_chunk`'s expected format, e.g.

````
|##@knitr foo
|... code ...
````

Then in the `.qmd` file:

````
```{r setup}
knitr::read_chunk('shared.R')
```
. . . . .

```{r foo}
```
````

to run the code.   I'm getting an error that a data table created inside `foo` is not found when a later R code piece in `shared.R` that uses that data table is invoked from `.qmd`.

Before submitting a bug report with a reproducible example is `Quarto` intended to do this, when doing a book?

**Comment by mcanouil** [comment]
Would it be possible to have a fully reproducible example?
I personally now have a repository with different issues as independent directory: https://github.com/mcanouil/quarto-issues
This makes thing easier for anyone to try and not to guess.

EDIT: Below a reproducible example for standalone HTML document (from https://bookdown.org/yihui/rmarkdown-cookbook/read-chunk.html) in which it works.
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/179794447-653f15b2-5289-4f6d-8923-b5f15d027f49.png">

<details><summary>See the code.</summary>

- `demo.qmd`
  ````md
  ---
  title: "Test"
  format: html
  ---

  Read an external script:

  ```{r}
  #| include: false
  knitr::read_chunk("shared.R")
  ```

  Now we can use the code, e.g.,

  ```{r test-a}
  #| echo: false
  ```

  ```{r test-b}
  #| fig.height: 4
  ```
  ````

- `shared.R`

  ```r
  ## ---- test-a --------
  1 + 1

  ## ---- test-b --------
  if (TRUE) {
    plot(cars)
  }
  ```

</details>

**Reply by cderv** [reply]
Yes I agree having an example would help. 

@harrelfe does it work in a single `.qmd` document ? As you are mentionning book chapter explicitly I am assuming yes ?

**Reply by mcanouil** [reply]
I made a book with my previous example and it worked.
Code at: https://github.com/mcanouil/quarto-issues/tree/main/quarto-cli-1458
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/179800299-8fc2fdb5-f49c-4030-9e9e-3cd489c1369d.png">

@harrelfe What's exactly in your R script? Are you using the same syntax defining chunks and their names as in my examples?

**Reply by harrelfe** [reply]
I can get your full example to work on my system but I'm still having problems with output from remote chunks actually appearing in my book chapter (for example a base graph rendered with a `plot()` method).  Here is an example remote chunk:

```
## ---- variogram --------

v <- Variogram(a, form=~ week | uid)
plot(v)
```

and the chapter's chunk where it is invoked:

````{r variogram,cap='Variogram, with assumed correlation pattern superimposed'}
#| label: fig-long-variogram
````

**Note**:: If I change the _chunk_ label (also inside `shared.R`) from variogram to fig-long-variogram I think it works.  Can't figure out why this would mattter.

**Note**: I set `knitr::set_alias(w = 'fig.width', h = 'fig.height', cap = 'fig.cap', scap ='fig.scap')` at the top.

**Reply by mcanouil** [reply]
Because the chunk code from the R script is referenced by label, so it has to be consistent between the R script chunk and the chunk call from within Quarto (or RMarkdown).

I do not know how the knitr::set_alias interact within Quarto and I can't reproduce since you only provide small partial pieces of your use case.
A complete and small reproducible example as the one I showed would help me/anyone help you more efficiently and precisely.

**Reply by harrelfe** [reply]
Sorry the way my examples originate lately is from converting large chapters from LaTeX or RMarkdown to Quarto which makes it sometimes hard to isolate problems.  The `alias` is not interacting with this problem so I should have just used `fig.cap` which would result in the same problem.  For now I'll assume that chunk labels need to be "big" e.g. `fig-chaptername-figurename` even though I would have expected `chaptername-figurename` to have worked (it didn't) exept that `#| label: fig-chaptername-figurename` is required.

**Reply by mcanouil** [reply]
I'll try later to see if can make something that reproduce this behaviour.

**Reply by harrelfe** [reply]
In running a lot of tests (but still being unable to put together a simple reproducible example) I have found the `read_chunk()` approach to be unreliable in `Quarto`.  Some scope is being lost, i.e., earlier chunks that produced new objects that I would have been sure were stored in the `.globalEnv` gave rise to later chunks not finding those new objects.

**Reply by mcanouil** [reply]
I should have a bit of time in the next few days to try to reproduce what you have.
Meantime if you don't have a small reproducible example, does your book happen to be on a public repository on GitHub?

**Reply by harrelfe** [reply]
It will be public on Github but not yet.

**Comment by jimjam-slam** [comment]
An alternative to `knitr::read_chunk` could be to use an [Include](https://quarto.org/docs/authoring/includes.html)!

I _think_ Quarto shortcodes are processed before computations happen, so you could theoretically either include the _content_ of a single chunk, like:

````
Blah blah blah...

```{r}
{{ < include script.r > }}
```
````

Or include one or more chunks as a `.qmd file:

````
Blah blah blah...

{{ < include script.qmd > }}

````

(You might need to confirm that my understanding is correct, though!)

**EDIT:** my bad! @cscheid [says Lua filters happen after knitr/jupyter.](https://twitter.com/scheidegger/status/1555405784481050625)

**Reply by cscheid** [reply]
@jimjam-slam so there's an extra wrinkle here.

_Lua filters_ happen after knitr/jupyter, but `{{< include >}}` is _not_ a lua filter. `include` happens to be specially implemented, exactly because we expect people to want to include computational content in their includes. I should have clarified, sorry!

**Reply by jimjam-slam** [reply]
Oh, awesome! Sounds like it could be a viable alternative, then! 😊
